dbeaver/dbeaver#21832 Add Cubrid Plugin Features

longhaseng52 opened the pull request:
- Driver download from maven repository
- Connection properties
- Create folders to categorized by user
- Display list of user
- Display table based on user
- Display system table
- Display view based on user
- Display system view

longhaseng52 committed:
Cubrid Plugin Extension
- Create Cubrid plugin
- Edit connection and driver properties

longhaseng52 committed:
Cubrid Plugin Extension
- display list of user
- create folders to catagorized by user

longhaseng52 committed:
Cubrid Plugin Extension
- display list of table
- display list of system table

longhaseng52 committed:
Cubrid Plugin Extension
- display list of view
- display list of system view

longhaseng52 committed:
Cubrid Plugin Extension
- Fixed create table with UI
- Fixed duplicated column
- Remove system table/view from other user
- Remove unimplemented procedure

longhaseng52 committed:
Cubrid Plugin Extension
- Fixed column not loading on server 11.0
- Fixed duplicated primary key, foreign key and index

LonwoLonwo commented:
Hello @longhaseng52 

Thanks for your contribution.

We will look at it.

kisoo-han commented:
@LonwoLonwo ,
Longha made this PR on behalf of the CUBRID TEAM.
We, CUBRID Corp, closely worked with Longha to write/review the codes.
You can view this PR as a CUBRID official contribution to DBeaver, not a qualification of individual contributor.

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid.ui/plugin.xml:
```diff
@@ -0,0 +1,51 @@
+
+    <extension point="org.jkiss.dbeaver.dataSourceView">
+        <view
+            class="org.jkiss.dbeaver.ext.generic.views.GenericConnectionPage"
```
Any specific reason to add info about the connection page here? Looks like no new database parameters for this page.

LonwoLonwo replied:
Or it is just for the name and icon?

longhaseng52 replied:
Yes, It is for changing the name.

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,17 @@
+Bundle-ActivationPolicy: lazy
+Bundle-RequiredExecutionEnvironment: JavaSE-17
+Bundle-Vendor: JKISS
+Import-Package: org.jkiss.dbeaver.tasks.nativetool,
```
Any specific reason to have here these two imports?

longhaseng52 replied:
It may be my mistake to import those two without intention. I will remove it.

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/CubridConstants.java:
```diff
@@ -0,0 +1,11 @@
+package org.jkiss.dbeaver.ext.cubrid;
```
```suggestion
/*
 * DBeaver - Universal Database Manager
 * Copyright (C) 2010-2023 DBeaver Corp and others
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.jkiss.dbeaver.ext.cubrid;
```

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/CubridDataSourceProvider.java:
```diff
@@ -0,0 +1,27 @@
+package org.jkiss.dbeaver.ext.cubrid;
```
```suggestion
/*
 * DBeaver - Universal Database Manager
 * Copyright (C) 2010-2023 DBeaver Corp and others
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.jkiss.dbeaver.ext.cubrid;
```

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/edit/CubridForeignKeyManager.java:
```diff
@@ -0,0 +1,27 @@
+package org.jkiss.dbeaver.ext.cubrid.edit;
```
```suggestion
/*
 * DBeaver - Universal Database Manager
 * Copyright (C) 2010-2023 DBeaver Corp and others
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.jkiss.dbeaver.ext.cubrid.edit;
```

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/edit/CubridIndexManager.java:
```diff
@@ -0,0 +1,27 @@
+package org.jkiss.dbeaver.ext.cubrid.edit;
```
```suggestion
/*
 * DBeaver - Universal Database Manager
 * Copyright (C) 2010-2023 DBeaver Corp and others
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.jkiss.dbeaver.ext.cubrid.edit;
```

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/edit/CubridPrimaryKeyManager.java:
```diff
@@ -0,0 +1,27 @@
+package org.jkiss.dbeaver.ext.cubrid.edit;
```
```suggestion
/*
 * DBeaver - Universal Database Manager
 * Copyright (C) 2010-2023 DBeaver Corp and others
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.jkiss.dbeaver.ext.cubrid.edit;
```

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/edit/CubridTableColumnManager.java:
```diff
@@ -0,0 +1,47 @@
+package org.jkiss.dbeaver.ext.cubrid.edit;
```
```suggestion
/*
 * DBeaver - Universal Database Manager
 * Copyright (C) 2010-2023 DBeaver Corp and others
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.jkiss.dbeaver.ext.cubrid.edit;
```

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/edit/CubridTableManager.java:
```diff
@@ -0,0 +1,58 @@
+package org.jkiss.dbeaver.ext.cubrid.edit;
```
```suggestion
/*
 * DBeaver - Universal Database Manager
 * Copyright (C) 2010-2023 DBeaver Corp and others
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.jkiss.dbeaver.ext.cubrid.edit;
```

LonwoLonwo commented:
Please add copyright for Java classes.
And reformat code for 4-space indentation.

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridSQLDialect.java:
```diff
@@ -0,0 +1,11 @@
+public class CubridSQLDialect extends GenericSQLDialect {
+	public static final String CUBRID_DIALECT_ID = "cubrid";
+
+	public CubridSQLDialect() {
```
Does Cubrid have specific keywords words? Or functions words. They can be added to this class.

longhaseng52 replied:
Currently, I don't have specific keywords or functions yet.

LonwoLonwo replied:
Ok

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridTableBase.java:
```diff
@@ -0,0 +1,111 @@
+			return container.getDataSource().getCubridUsers(null);
+		} catch (DBException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
```
Please add logging instead.
Like here https://github.com/dbeaver/dbeaver/blob/57dd446701282f25071aec4df9b86a387d74e772/plugins/org.jkiss.dbeaver.ext.postgresql/src/org/jkiss/dbeaver/ext/postgresql/model/PostgreRole.java#L358

With a clear message.

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridTableBase.java:
```diff
@@ -0,0 +1,111 @@
+	}
+
+	public Collection<? extends CubridTableIndex> getIndexes(DBRProgressMonitor monitor) throws DBException {
+		if (getDataSource().getInfo().supportsIndexes()) {
```
Why do you have such a condition here? Does Cubrid support indexes? If it supports indexes - then just return the cache.

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridTableBase.java:
```diff
@@ -0,0 +1,111 @@
+	
+	private CubridUser owner;
+	private CubridObjectContainer container;
+	private String isSystemTable;
```
Let's use a boolean variable instead.
And please implement the DBPSystemObject interface for this value.

longhaseng52 replied:
- I have changed it to boolean type.
- DBPSystemObject: is already Implemented in GenericTableBase.

LonwoLonwo replied:
Then you can override this method:

```java
@Override
    public boolean isSystem() {
        return isSystemTable;
    }
```

longhaseng52 replied:
This will hide the table when it true, but my scenario is to separate query between system-table and user-table, and keep all table to display.

LonwoLonwo replied:
> This will hide the table when it true

![2023-12-11 15_01_23-DBeaver Ultimate 23 3 0 - _duck_090 db_ Script-144](https://github.com/dbeaver/dbeaver/assets/45152336/33818c82-8e91-42fe-9d5b-7dbc752aef52)

Yes, and it is expected. And it is customizing this way.
Why do you want not to hide system tables?

longhaseng52 replied:
Thank you for your suggestion. I didn't notice this feature before. 
Now, I have changed the code to follow your suggestion.

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridTableBase.java:
```diff
@@ -0,0 +1,111 @@
+		if(isSystemTable.equals("YES")) {
+			return this.getName();
+		} else {
+			return this.getOwner().getName() + "." + this.getName();
```
Please use 
`DBUtils.getFullQualifiedName(getDataSource(), getOwner(), this);`
instead. Because you do not have a quoting condition here.

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridTableBase.java:
```diff
@@ -0,0 +1,111 @@
+	@Override
+	public String getFullyQualifiedName(DBPEvaluationContext context) {
+		if(isSystemTable.equals("YES")) {
+			return this.getName();
```
Please use
DBUtils.getFullQualifiedName(getDataSource(), this);
instead. Because you do not have a quoting condition here.

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/meta/CubridMetaModel.java:
```diff
@@ -0,0 +1,174 @@
+		return dbStat;
+	}
+
+	public boolean isNewVersion(JDBCSession session) {
```
You can't use here "new version" name here. It is new now, but you need to think about the future in the present. A few years later this version will be not new, but this method will have such a name.

So, please change it to "isDatabaseVersionAtLeast11And2" or something like this.

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/meta/CubridMetaModel.java:
```diff
@@ -0,0 +1,174 @@
+				return true;
+			}
+		} catch (SQLException e) {
+			e.printStackTrace();
```
Please use logging here with a clear message.

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/meta/CubridMetaModelDescriptor.java:
```diff
@@ -0,0 +1,112 @@
+import org.jkiss.utils.ArrayUtils;
+import org.jkiss.utils.CommonUtils;
+
+public class CubridMetaModelDescriptor extends AbstractDescriptor {
```
I do not see any reason for the existence of this class. Our GenericMetaModelDescriptor is working well. And you have all the necessary ID-s in your plugin.xml. So, please remove it.

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridObjectContainer.java:
```diff
@@ -0,0 +1,324 @@
+		@NotNull
+		@Override
+		protected JDBCStatement prepareObjectsStatement(@NotNull JDBCSession session, @NotNull CubridObjectContainer container) throws SQLException {
+			return container.getDataSource().getMetaModel().prepareCubridUserLoadStatement(session, container);
```
No need to do it in the meta-model class. Please move your code here.

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridObjectContainer.java:
```diff
@@ -0,0 +1,324 @@
+		protected CubridUser fetchObject(JDBCSession session, CubridObjectContainer container, JDBCResultSet resultSet) throws SQLException, DBException {
+			String name = resultSet.getString("name");
+			String comment = resultSet.getString("comment");
+			return container.getDataSource().getMetaModel().createCubridUserImpl(container, name, comment);
```
No need to do it in the meta-model class. Please move your code here.

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridObjectContainer.java:
```diff
@@ -0,0 +1,324 @@
+		@Override
+		public @NotNull JDBCStatement prepareLookupStatement(@NotNull JDBCSession session, @NotNull CubridObjectContainer owner,
+			@Nullable CubridTableBase object, @Nullable String objectName) throws SQLException {
+			return dataSource.getMetaModel().prepareSystemTableLoadStatement(session, owner, object, objectName);
```
No need to do it in the meta-model class. Please move your code here.

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridObjectContainer.java:
```diff
@@ -0,0 +1,324 @@
+		@Override
+		public @NotNull JDBCStatement prepareLookupStatement(@NotNull JDBCSession session, @NotNull CubridObjectContainer owner,
+			@Nullable CubridTableBase object, @Nullable String objectName) throws SQLException {
+			return dataSource.getMetaModel().prepareSystemViewLoadStatement(session, owner, object, objectName);
```
No need to do it in the meta-model class. Please move your code here.

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridObjectContainer.java:
```diff
@@ -0,0 +1,324 @@
+
+	}
+
+	public class CubridIndexCache extends JDBCCompositeCache<CubridObjectContainer, CubridTableBase, CubridTableIndex, GenericTableIndexColumn>{
```
I do not see any specific reason for CubridTableIndex to exist. Why can't you just use the Generic index?

longhaseng52 replied:
I want to override **prepareObjectsStatement** to support separate index based on owner. but generic's indexcache is not allow to override it from meta-model class.

LonwoLonwo replied:
Ok.

longhaseng52 committed:
Fixed Issues
- removed unused imported packages
- added copyright for java classes
- added logs
- optimized codes
- naming convention

longhaseng52 committed:
Fix Issues
- Override isSystem() & customize display of system table.
- Optimized code to reflect Codacy suggestions

Destrolaric commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridDataSource.java:
```diff
@@ -0,0 +1,60 @@
+		return metaModel;
+	}
+    
+	@Override
```
Small formatting issue

Destrolaric commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridView.java:
```diff
@@ -0,0 +1,45 @@
+import org.jkiss.dbeaver.model.exec.jdbc.JDBCResultSet;
+import org.jkiss.dbeaver.model.struct.DBSObjectWithScript;
+
+public class CubridView extends CubridTableBase implements DBSObjectWithScript {
```
Please add `implements DBSView`. And also, is it possible to inherit this from GenericView without significant code changes and functionality duplication?

longhaseng52 replied:
Noted, I will implements **DBSView**. 
I extends  **CubridTableBase**  because it shares some features to **CubridView** like: **Owner** ... etc.

Destrolaric commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridTableForeignKey.java:
```diff
@@ -0,0 +1,32 @@
+
+public class CubridTableForeignKey extends GenericTableForeignKey {
+
+	public CubridTableForeignKey(CubridTable table, String name, String remarks, DBSEntityReferrer referencedKey,
+		DBSForeignKeyModifyRule deleteRule, DBSForeignKeyModifyRule updateRule,
+		DBSForeignKeyDeferability deferability, boolean persisted) {
+		super(table, name, remarks, referencedKey, deleteRule, updateRule, deferability, persisted);
+	}
```
Is CubridTableForeignKey needed, it looks like we can just use GenericTableForeignKey

longhaseng52 replied:
Yes, It needed. If I use `GenericTableForeignKey`, `CubridForeignKeyManager` will be overridden by the `GenericForeignKeyManager` in configuration `plugin.xml` file.

Imagine if I use GenericTableForeignKey, the extension of Manager will be:

    <manager class="org.jkiss.dbeaver.ext.cubrid.edit.CubridForeignKeyManager" objectType="org.jkiss.dbeaver.ext.generic.model.GenericTableForeignKey"/>

It will override by generic plugin.xml when build.

    <manager class="org.jkiss.dbeaver.ext.generic.edit.GenericForeignKeyManager"    objectType="org.jkiss.dbeaver.ext.generic.model.GenericTableForeignKey"/>

Please let me know, if you have any solutions for this problem.

Destrolaric replied:
We can try to use GenericTableForeignKey. It looks like there is no significant difference in constructors

longhaseng52 replied:
I have fixed it by removing CubridTableForeignKey and using GenericTableForeignKey instead.

Destrolaric commented on plugins/org.jkiss.dbeaver.ext.cubrid/plugin.xml:
```diff
@@ -0,0 +1,171 @@
+                        id="cubrid_jdbc"
+                        label="CUBRID"
+                        icon="icons/cubrid_icon.png"
+                        iconBig="icons/cubrid_icon_big.png"
+                        class="cubrid.jdbc.driver.CUBRIDDriver"
```
You can also add logoImage icon here, this will allow logo to be shown in connection settings without the need to override generic connection page

longhaseng52 replied:
I have fixed it and removed Connection page configuration.

Destrolaric commented on plugins/org.jkiss.dbeaver.ext.cubrid/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,15 @@
+Require-Bundle: org.eclipse.core.runtime,
+ org.jkiss.dbeaver.model;visibility:=reexport,
+ org.jkiss.dbeaver.model.sql;visibility:=reexport,
+ com.google.guava;resolution:=optional,
```
I don't see any direct guava usage in the classes. Is this dependency required?

longhaseng52 replied:
I have fixed it.

Destrolaric commented on plugins/org.jkiss.dbeaver.ext.cubrid.ui/plugin.xml:
```diff
@@ -0,0 +1,51 @@
+        </view>
+    </extension>
+
+    <extension point="org.jkiss.dbeaver.databaseEditor">
+        <editor
+            id="cubrid.table.ddl.view"
+            class="org.jkiss.dbeaver.ui.editors.sql.SQLSourceViewer"
+            main="false"
+            label="DDL"
+            description="DDL"
+            position="additions_end"
+            icon="#sql_text"
+            type="folder">
+            <objectType name="org.jkiss.dbeaver.ext.cubrid.model.CubridTable"/>
+        </editor>
+        <editor
+            id="cubrid.view.source.view"
+            class="org.jkiss.dbeaver.ext.generic.editors.GenericSourceViewEditor"
+            main="false"
+            label="Definition"
+            description="View definition"
+            position="additions_end"
+            icon="#sql_text"
+            type="folder"
+            embeddable="false">
+            <objectType name="org.jkiss.dbeaver.ext.cubrid.model.CubridView"/>
+        </editor>
+    </extension>
+
```
Is it possible to change the logic to use GenericView implementation instead?

longhaseng52 replied:
I have already fixed it. It's no longer need View DDL configuration anymore. So I removed it.

Destrolaric commented on plugins/org.jkiss.dbeaver.ext.cubrid.ui/plugin.xml:
```diff
@@ -0,0 +1,51 @@
+            label="DDL"
+            description="DDL"
+            position="additions_end"
+            icon="#sql_text"
+            type="folder">
+            <objectType name="org.jkiss.dbeaver.ext.cubrid.model.CubridTable"/>
+        </editor>
```
Not sure we want to declare it for DDL viewer again, we can use the definition `generic.table.ddl.view` if condition `object.supportsDDL()` is true.

longhaseng52 replied:
The DDL still not showing when I don't declare it again even `object.supportsDDL()` return true. I think it's because of  `<objectType>`. Since I use `CubridTable` but generic use `GenericTable`.

Destrolaric commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/meta/CubridMetaModel.java:
```diff
@@ -0,0 +1,133 @@
+	}
+
+	public boolean isSupportMultiSchema(JDBCSession session) {
+		try {
+			int major = session.getMetaData().getDatabaseMajorVersion();
+			int minor = session.getMetaData().getDatabaseMinorVersion();
+			if(major == 11 && minor >= 2) {
+				return true;
+			}else if(major > 11) {
+				return true;
+			}
+		} catch (SQLException e) {
+			log.error("can't get database version");
+		}
```
Formatting issues and exception should passed as `log.error` argument

Destrolaric replied:
Also you conditions can be concatenated and be `major > 11 || (major == 11 && minor >= 2)` instead of two similar if conditions

longhaseng52 replied:
I have fixed it.

Destrolaric commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridTableBase.java:
```diff
@@ -0,0 +1,138 @@
+		try {
+			return container.getDataSource().getCubridUsers(null);
+		} catch (DBException e) {
+			log.error("Cannot get user.");
```
It's better to add an exception to `log.error` when caught

longhaseng52 replied:
I have fixed it.

Destrolaric commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridUniqueKey.java:
```diff
@@ -0,0 +1,29 @@
+
+public class CubridUniqueKey extends GenericUniqueKey{
+
+	public CubridUniqueKey(CubridTable table, String name, String remarks, DBSEntityConstraintType constraintType,
+		boolean persisted) {
+		super(table, name, remarks, constraintType, persisted);
+	}
```
Maybe we can use generic implementation directly?

longhaseng52 replied:
For `CubridUniqueKey`, it is same case as `CubridForeignKey`. Please kindly refer to my comment on `CubridTableForeignKey` above.

longhaseng52 replied:
I have fixed it by using generic implementation directly.

Destrolaric commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridTableBase.java:
```diff
@@ -0,0 +1,138 @@
+
+public abstract class CubridTableBase extends GenericTableBase{
+	
+    private static final Log log = Log.getLog(CubridTableBase.class);
+
+	private CubridUser owner;
+	private CubridObjectContainer container;
+	private boolean isSystemTable;
```
Something happened with formatting here

longhaseng52 replied:
I have fixed it.

Destrolaric commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/CubridConstants.java:
```diff
@@ -0,0 +1,27 @@
+
+import org.jkiss.dbeaver.ext.generic.GenericConstants;
+
+public class CubridConstants extends GenericConstants {
+	
+	public static final String OWNER_NAME = "owner_name";
+	public static final String CLASS_NAME = "class_name";
+	public static final String IS_SYSTEM_CLASS = "is_system_class";
```
I think it's better not to extend GenericConstants but to use them directly when required.

longhaseng52 replied:
I have fixed it.

Destrolaric commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridDataSource.java:
```diff
@@ -0,0 +1,60 @@
+	private final CubridMetaModel metaModel;
+	private CubridObjectContainer structureContainer;
+	
+	public CubridDataSource(DBRProgressMonitor monitor, DBPDataSourceContainer container, CubridMetaModel metaModel) throws DBException {
+		super(monitor, container, metaModel, new CubridSQLDialect());
+		this.metaModel = new CubridMetaModel();
+    }
```
Formatting

longhaseng52 replied:
I have fixed it.

Destrolaric reviewed (changes requested):
Greentings, thanks for the pull request. Can you look at my comments mentioning formatting issues and logging? Also, some questions about the configuration

longhaseng52 committed:
Change CubridTable & CubridView structure implementation

longhaseng52 committed:
Create CubridViewManager & Implement DBSView in CubridView

longhaseng52 committed:
Remove CubridTableForeignkey & CubridForeignkeyManager

longhaseng52 committed:
Remove CubridUniquekey & CubridPrimarykeyManager

longhaseng52 committed:
Remove CubridTableIndex & CubridIndexManager

longhaseng52 committed:
Remove Cubrid UI plugin
- Remove DDL declaration
- Remove Cubrid connection page configuration

longhaseng52 committed:
Concatenated condition in isSupportMultiSchema()

longhaseng52 committed:
Remove unused dependency

longhaseng52 committed:
Remove GenericConstants from CubridConstants

longhaseng52 committed:
Change Cubrid User's icon

longhaseng52 committed:
Add an exception to log.error

longhaseng52 committed:
Formatting 4-space indentation

longhaseng52 committed:
Remove Cubrid UI from pom.xml and feature.xml

longhaseng52 committed:
Merge remote-tracking branch 'origin/develop' into 1st_pr_official
Resolved merge conflict by keeping the latest feature.xml format

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridView.java:
```diff
@@ -0,0 +1,52 @@
+	@Override
+	@Property(hidden = true, editable = true, updatable = true, order = -1)
+	public String getObjectDefinitionText(DBRProgressMonitor monitor, Map<String, Object> options) throws DBException {
+		return getDataSource().getMetaModel().getViewDDL(monitor, null, options);
```
Does Cubrid have information about view definitions anywhere?

longhaseng52 replied:
It doesn't have  yet. We plan to add this feature in the next coming PR.

longhaseng52 committed:
Fixed Codacy Static Code Analysis Suggestion

longhaseng52 committed:
Fixed Resource 'meta.org.jkiss.dbeaver.ext.cubrid.model.CubridTable.owner.name' not found in bundle

Destrolaric commented on plugins/org.jkiss.dbeaver.ext.cubrid/plugin.xml:
```diff
@@ -0,0 +1,170 @@
+
+<plugin>
+  
+  	<extension point="org.jkiss.dbeaver.generic.meta">
+        <meta id="cubrid" class="org.jkiss.dbeaver.ext.cubrid.model.meta.CubridMetaModel" driverClass="cubrid.jdbc.driver.CUBRIDDriver"/>
+    </extension>
```
Formatting

longhaseng52 replied:
I have fixed it.

Destrolaric commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/edit/CubridViewManager.java:
```diff
@@ -0,0 +1,28 @@
+public class CubridViewManager extends GenericViewManager {
+
+	@Override
+	public boolean canCreateObject(Object container) {
+		return false;
+	}
```
Are views somehow created automatically in Cubrid, or the driver doesn't support creation?

longhaseng52 replied:
Cubrid support view creation.
I think the View creation is really similar to when we create view on SQL editor. So I disabled it and let user create view from SQL editor instead.
If you think it should be supported the View creation, I can enable it.

Destrolaric commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridObjectContainer.java:
```diff
@@ -0,0 +1,360 @@
+		return tables;
+	}
+
+	public List<? extends CubridView> getViews(DBRProgressMonitor monitor, String name) throws DBException {
+		List<CubridView> tables = new ArrayList<>();
+		for (CubridTable table : this.cubridTableCache.getAllObjects(monitor, this)) {
+			if (table.isView() && table.getOwner().getName().equals(name)) {
+				tables.add((CubridView) table);
+			}
+		}
+		return tables;
+	}
+
+	public List<? extends CubridView> getSystemViews(DBRProgressMonitor monitor, String name) throws DBException {
+		List<CubridView> views = new ArrayList<>();
+		for (CubridTable table : this.cubridSystemViewCache.getAllObjects(monitor, this)) {
+			if (table.getOwner().getName().equals(name)) {
+				views.add((CubridView) table);
+			}
+		}
+		return views;
+	}
```
Maybe caching will be better?

longhaseng52 replied:
I have fixed it.

Destrolaric commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridTable.java:
```diff
@@ -0,0 +1,178 @@
+
+	public CubridTable(CubridObjectContainer container, @Nullable String tableName, @Nullable String tableType,
+			@Nullable JDBCResultSet dbResult) {
+		super(container, tableName, tableType, dbResult);
+		this.container = container;
```
I think the container was already set in super constructor

longhaseng52 replied:
I have fixed it.

Destrolaric commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/meta/CubridMetaModel.java:
```diff
@@ -0,0 +1,120 @@
+		String tableName = JDBCUtils.safeGetString(dbResult, CubridConstants.CLASS_NAME);
+		String tableType = JDBCUtils.safeGetStringTrimmed(dbResult, JDBCConstants.TABLE_TYPE);
+
+		CubridTable table = this.createTableImpl(owner, tableName, tableType, dbResult);
```
You can pass session monitor to createTableImpl, so the function may know if loading was cancelled by user

longhaseng52 replied:
I have fixed it.

Destrolaric commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridTable.java:
```diff
@@ -0,0 +1,178 @@
+
+	public Collection<? extends CubridUser> getUsers() {
+		try {
+			return container.getDataSource().getCubridUsers(null);
```
Avoid using null in monitor parameters better to use
`new VoidProgressMonitor()`, but I would recommend reading my comments about passing the session monitor to properly handle loading any metadata

longhaseng52 replied:
I have fixed it.

Destrolaric commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridTable.java:
```diff
@@ -0,0 +1,178 @@
+			isSystemTable = false;
+		}
+
+		for (CubridUser cbOwner : getUsers()) {
+			if (cbOwner.getName().equals(owner_name)) {
+				this.owner = cbOwner;
+				break;
+			}
+		}
```
If you require to load metadata in constructor you can pass session monitor parameter to constructor, so the loading can be properly monitored in session

longhaseng52 replied:
I have fixed it.

Destrolaric commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridTable.java:
```diff
@@ -0,0 +1,178 @@
+			isSystemTable = false;
+		}
+
+		for (CubridUser cbOwner : getUsers()) {
+			if (cbOwner.getName().equals(owner_name)) {
+				this.owner = cbOwner;
+				break;
+			}
+		}
```
You can pass JDBCSession.getProgressMonitor(), so the loading operation for users will be cancellable, see `@NotNull JDBCSession session` in `createTableImpl`

longhaseng52 replied:
I have fixed it.

Destrolaric commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/model/CubridTable.java:
```diff
@@ -0,0 +1,178 @@
+		return this.container;
+	}
+
+	public Collection<? extends CubridUser> getUsers() {
```
Add monitor parameter here

longhaseng52 replied:
I have fixed it.

Destrolaric reviewed (commented):
Hi, it looks like Java files are indented with TABs could you please reindent them with 4 spaces? I also added some suggestions below

longhaseng52 committed:
Added session monitor parameter in CubridTable

longhaseng52 committed:
Unassigned container in CubridTable's constructor

longhaseng52 committed:
Fixed system view caching

longhaseng52 committed:
Fixed indentation in plugin.xml file

longhaseng52 committed:
Fixed null container in CubridTable

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/plugin.xml:
```diff
@@ -0,0 +1,170 @@
+                                                </items>
+                                            </items>
+                                        </folder>
+                                        <folder label="%tree.references.node.name" icon="#references" description="%tree.references.node.tip" visibleIf="!object.view &amp;&amp; object.dataSource.info.supportsReferentialIntegrity()" virtual="true">
```
```suggestion
                                        <folder label="%tree.references.node.name" icon="#references" description="%tree.references.node.tip" virtual="true">

```

LonwoLonwo replied:
You can remove this condition because you already know that your database supports it.

longhaseng52 replied:
I have removed it.

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/plugin.xml:
```diff
@@ -0,0 +1,170 @@
+                                            </items>
+                                        </items>
+                                    </folder>
+                                    <folder label="%tree.references.node.name" icon="#references" description="%tree.references.node.tip" visibleIf="!object.view &amp;&amp; object.dataSource.info.supportsReferentialIntegrity()" virtual="true">
```
```suggestion
                                    <folder label="%tree.references.node.name" icon="#references" description="%tree.references.node.tip" virtual="true">

```

longhaseng52 replied:
I have removed it.

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/plugin.xml:
```diff
@@ -0,0 +1,170 @@
+           	                        </items>
+           	                    </folder>
+                            </folder>
+                            <folder type="org.jkiss.dbeaver.ext.cubrid.model.CubridView" label="%tree.tviews.node.name" icon="#folder_view" description="%tree.tviews.node.tip" visibleIf="object.dataSource.metaModel.supportsViews(object.dataSource)">
```
```suggestion
                            <folder type="org.jkiss.dbeaver.ext.cubrid.model.CubridView" label="%tree.tviews.node.name" icon="#folder_view" description="%tree.tviews.node.tip">

```

longhaseng52 replied:
I have removed it.

LonwoLonwo commented on plugins/org.jkiss.dbeaver.ext.cubrid/src/org/jkiss/dbeaver/ext/cubrid/CubridConstants.java:
```diff
@@ -0,0 +1,25 @@
+/*
+ * DBeaver - Universal Database Manager
+ * Copyright (C) 2010-2023 DBeaver Corp and others
```
```suggestion
 * Copyright (C) 2010-2024 DBeaver Corp and others

```

LonwoLonwo replied:
Please update the year in the copyrights.

longhaseng52 replied:
I have updated it.

LonwoLonwo commented:
Please start your commits with the "# + ticket id", like:

`#21832 Update plugin.xml file`

It will help us in the future to find the original ticket from the code annotations.

LonwoLonwo committed:
Merge branch 'devel' into 1st_pr_official

longhaseng52 committed:
#21832 Update plugin.xml file

longhaseng52 committed:
#21832 Update the year in the copyrights

LonwoLonwo commented:
Hello again @longhaseng52 

We discussed your PR with our team, and unfortunately, we will ask you to add more changes.

And first of all - formatting. Still, it is hard to read your code. Please reindent your JAVA files with 4 spaces.

Second - could you please provide your navigator metadata model? I see in the plugin.xml file - databases -> schemas -> users -> tables. And then again, users. Why is it so complicated? Does it work like in Oracle, where user == schema?

Third. Managers. Could you please explain the special meanings of the CubridTableColumnManager, CubridTableManager, and CubridViewManager? I see only copy-paste code from the generic part. As I said before, we are training to avoid copy-paste code in our bundles. If you want to create your objects, you can use methods like getDataSource().getMetaModel().createTableImpl.

kisoo-han commented:
Hello @LonwoLonwo 

For your change requests,
1. We will modify java indent according to your recommendation (4 spaces).
2. Basically, CUBRID's meta data model is almost the same as that of ORACLE (schema = user). Considering the perspective of a consistent UI view of DBeaver product, we will change the navigator to be simple and intuitive.
Following is the meta data model of CUBRID, we plan to add code for processing the remaining meta data in the following PR (* means it is not included in the 1st PR).

(database)->{schema, users, data types}
{schema}->{table, view, index*, serial*, procedure*, function*, synonym*, trigger*}

3. We planned a table/column/view manager specialized for CUBRID. To reduce the complexity of the first PR, we have removed many planned features. As a result, and as you pointed out, managers could be replaced by the generic in current implementation. So, we will remove them and when we need to put additional features in managers, we will create them (2nd PR).

LonwoLonwo commented:
Hello @kisoo-han 

Ok, that sounds like a good plan (about the second PR). And thanks for the explanations.

longhaseng52 committed:
#21832 formatting code with 4 spaces

longhaseng52 committed:
#21832 change navigator structure in plugin.xml

longhaseng52 committed:
#21832 restructure schema

longhaseng52 committed:
#21832 remove cubrid manager class

longhaseng52 committed:
#21832 remove duplicated data type

longhaseng52 committed:
#21832 fixed Codacy Static Code Analysis suggestion

LonwoLonwo commented:
Thanks for the answer and for the changes.

uslss committed:
Merge branch 'devel' into 1st_pr_official
